File: LWWebSocket/Classes/WebSocketManager.m
//
//  WebSocketManager.m
//  YYMobileCore
//
//  Created by liusilan on 2017/4/7.
//  Copyright © 2017年 YY.inc. All rights reserved.
//

#import "WebSocketManager.h"
#import "HTTPConnection.h"
#import "HTTPServer.h"
#import "WebSocket.h"



@class MyWebSocket;

@interface WebSocketManager ()<WebSocketDelegate>
@property (nonatomic, strong) HTTPServer *httpServer;
@property (nonatomic, weak) MyWebSocket *webSocket;

@property(nonatomic, strong) NSOutputStream *dataStream;
@property(nonatomic, copy) NSString *streamFilePath;
@property(nonatomic, strong) NSError *streamError;
- (NSString *)myURI;
@end



@interface MyWebSocket : WebSocket
@property(nonatomic, copy) NSString *uri;
- (void)startHeartBeatRecvTimer;
@end

@implementation MyWebSocket{
    dispatch_source_t _heartBeatRecvTimer;
    dispatch_queue_t _heartBeatRecvQueue;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - heartBeat

// 心跳回包计时器
- (void)startHeartBeatRecvTimer {
    [self stopHeartBeatRecvTimer];

    __weak typeof(self) wself = self;
    if (!_heartBeatRecvQueue) {
        _heartBeatRecvQueue = dispatch_queue_create("com.wodedata.mobile.heatBeatRecv", DISPATCH_QUEUE_SERIAL);
    }

    _heartBeatRecvTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, _heartBeatRecvQueue);
    dispatch_source_set_timer(_heartBeatRecvTimer, dispatch_walltime(NULL, 8 * NSEC_PER_SEC), DISPATCH_TIME_FOREVER, (1ull * NSEC_PER_SEC) / 10);

    dispatch_source_set_event_handler(_heartBeatRecvTimer, ^{
        [wself timerHandle];
    });
    dispatch_resume(_heartBeatRecvTimer);
}

- (void)stopHeartBeatRecvTimer {

    if (_heartBeatRecvTimer) {
        dispatch_source_cancel(_heartBeatRecvTimer);
        _heartBeatRecvTimer = nil;
    }
}

// 超时未收到回包，关闭连接
- (void)timerHandle {
    [self stopHeartBeatRecvTimer];

    [self stop];
}

@end



#pragma mark - MyHTTPConnection

@interface MyHTTPConnection : HTTPConnection {
}
@property(nonatomic, strong) MyWebSocket *webSocket;
@end

@implementation MyHTTPConnection

- (WebSocket *)webSocketForURI:(NSString *)path {
    NSString *myURI = WebSocketManager.sharedManager.myURI;
    if ([path isEqualToString:myURI]) {
        self.webSocket = [[MyWebSocket alloc] initWithRequest:request socket:asyncSocket];
        self.webSocket.uri = path;
        self.webSocket.delegate = WebSocketManager.sharedManager;
        return self.webSocket;
    }

    return [super webSocketForURI:path];
}

@end




#pragma mark - WebSocketManager

@implementation WebSocketManager

+ (instancetype)sharedManager {
    static WebSocketManager *manager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        manager = [WebSocketManager new];
    });
    
    return manager;
}

-(NSString *)streamFilePath {
    if(!_streamFilePath){
        _streamFilePath = [NSTemporaryDirectory() stringByAppendingPathComponent:NSUUID.UUID.UUIDString];
    }
    return _streamFilePath;
}

-(NSOutputStream *)dataStream {
    if(!_dataStream){
        _dataStream = [[NSOutputStream alloc] initToFileAtPath:self.streamFilePath append:YES];
    }
    return _dataStream;
}

- (void)startServerWithPort:(UInt16)port webPath:(NSString *)webPath {
    
    if (_httpServer.isRunning) {
        return;
    }
    
    _httpServer = [[HTTPServer alloc] init];
    
    // Tell server to use our custom MyHTTPConnection class.
    [_httpServer setConnectionClass:[MyHTTPConnection class] ];

    // Tell the server to broadcast its presence via Bonjour.
    // This allows browsers such as Safari to automatically discover our service.
    [_httpServer setType:@"_http._tcp."];
    
    // Normally there's no need to run our server on any specific port.
    // Technologies like Bonjour allow clients to dynamically discover the server's port at runtime.
    // However, for easy testing you may want force a certain port so you can just hit the refresh button.
    [_httpServer setPort:port];

    //NSString *webPath = NSHomeDirectory();
    NSString *documentPath = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES).firstObject;
    NSString *absWebPath = documentPath;
    if(![webPath hasPrefix:@"/var"]){
        absWebPath = [documentPath stringByAppendingPathComponent:webPath];
    }else{
        absWebPath = webPath;
    }
    BOOL isDirectory = NO;
    BOOL existWebPath = [NSFileManager.defaultManager fileExistsAtPath:absWebPath isDirectory:&isDirectory];
    if(!isDirectory || !existWebPath) { //文件夹不存在
        NSError *error;
        BOOL success = [NSFileManager.defaultManager createDirectoryAtPath:absWebPath withIntermediateDirectories:YES attributes:nil error:&error];
        WSLog(@"success = %d，error = %@", success,error);
        absWebPath = success ? absWebPath : documentPath;
    }

    [_httpServer setDocumentRoot:absWebPath];

    WSLog(@"=====webpath:%@",webPath);

    // Start the server (and check for problems)
    
    NSError *error;
    if(![_httpServer start:&error]) {
        WSLog(@"start server error:%@", error.localizedDescription);
    }
}

- (void)stopServer {

    if (_httpServer && _httpServer.isRunning) {
        [_httpServer stop];
    }
}

- (NSString *)myURI {
    return @"/service";
}

-(BOOL)sendMessage:(NSString *)message {
    if(!self.webSocket){
        return NO;
    }

    [self.webSocket sendMessage:message];

    return YES;
}

-(BOOL)sendData:(NSData *)data {
    if(!self.webSocket){
        return NO;
    }

    NSMutableData *sendData = [self constructDataWithMessageType:SocketMessageType_Data];
    [sendData appendData:data];
    [self.webSocket sendData:sendData isBinary:YES];

    return YES;
}

-(BOOL)sendDataWithFileURL:(NSURL *)fileURL {
    if(!self.webSocket){
        return NO;
    }

    NSMutableData *startData = [self constructDataWithMessageType:SocketMessageType_StreamStart];
    [self.webSocket sendData:startData isBinary:YES];

    NSError *error;
    NSFileHandle * fileHandle = [NSFileHandle fileHandleForReadingFromURL:fileURL error:&error];
    if(error){
        WSLog(@"===error:%@",error.localizedDescription);
    }
    NSData * data = nil;
    while ((data = [fileHandle readDataOfLength:10240])) {
        if(data.length > 0){
            NSMutableData *streamingData = [self constructDataWithMessageType:SocketMessageType_Streaming];
            [streamingData appendData:data];
            [self.webSocket sendData:streamingData isBinary:YES];
        }else{
            break;
        }
    }
/*
    NSInputStream *inputStream = [[NSInputStream alloc] initWithFileAtPath:filePath];
    uint8_t buffer[10240];
    int len = 0;
    while ([inputStream hasBytesAvailable]) {
        len = [inputStream read:buffer maxLength:sizeof(buffer)];
        if (len > 0) {
            NSMutableData *streamingData = [self constructDataWithMessageType:SocketMessageType_Streaming];
            [streamingData appendBytes:buffer length:(NSUInteger)len];
            [self.webSocket sendData:streamingData isBinary:YES];
        }
    }
*/

    NSMutableData *endData = [self constructDataWithMessageType:SocketMessageType_StreamEnd];
    [self.webSocket sendData:endData isBinary:YES];

    return YES;
}


//构造数据头部
- (NSMutableData *)constructDataWithMessageType:(uint32_t)messageType {
    NSMutableData *data = [NSMutableData dataWithCapacity:0];
    [data appendBytes:&messageType length:sizeof(uint32_t)];
    return data;
}


//发送第一个活跃状态的二进制测试消息
- (void)sendActiveBinaryData {
    NSMutableData *data = [NSMutableData dataWithCapacity:0];

    uint32_t messageType = SocketMessageType_Hello;
    [data appendBytes:&messageType length:sizeof(uint32_t)];

    [data appendData:[@"hello world!" dataUsingEncoding:NSUTF8StringEncoding]];

    [self.webSocket sendData:data isBinary:YES];
}



#pragma mark - WebSocketDelegate

//连接打开
- (void)webSocketDidOpen:(WebSocket *)ws {
    WSLog(@"=======%s", __FUNCTION__);

    if([ws isKindOfClass:[MyWebSocket class]]){
        self.webSocket = (MyWebSocket *)ws;
    }
    [self.webSocket startHeartBeatRecvTimer];

    [self sendActiveBinaryData];
}

//收到消息串
- (void)webSocket:(WebSocket *)ws didReceiveMessage:(NSString *)msg{
    WSLog(@"=======%s", __FUNCTION__);
    MyWebSocket *webSocket = (MyWebSocket *)ws;

    if (msg.length <= 0) {
        return;
    }

    WSLog(@"=======didReceiveMessage msg:%@", msg);

    NSData *data = [msg dataUsingEncoding:NSUTF8StringEncoding];
    if (!data) {
        return;
    }

    NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:nil];
    uint32_t messageType = (uint32_t)[dict[@"messageType"] integerValue];
    if(messageType == SocketMessageType_HeartBeat){
        [webSocket startHeartBeatRecvTimer];
        return;
    }

    NSString *messageBody = dict[@"messageBody"];

    //处理msg
    if(self.handleReceiveMessage){
        self.handleReceiveMessage(messageType,messageBody);
    }

/*
    NSString *messageType = dict[@"messageType"];
    if (messageType.length <= 0) {
        return;
    }

    if ([messageType isEqualToString:@"heartBeat"]) {
        [webSocket startHeartBeatRecvTimer];
        return;
    }
*/

}

//收到二进制数据
- (void)webSocket:(WebSocket *)ws didReceiveData:(NSData *)data{
    WSLog(@"=======%s",__FUNCTION__);
    MyWebSocket *webSocket = (MyWebSocket *)ws;
    if(!self.handleReceiveData){
        return;
    }

    NSUInteger headerLen = 4;
    if(data.length < headerLen){
        return; //消息不合法
    }

    NSData *headerData = [data subdataWithRange:NSMakeRange(0, headerLen)];
    if(headerData.length <= 0 ){
        return; //获取消息错误
    }

    //appid，截取前4个字节
    uint32_t messageType;
    [headerData getBytes:&messageType length:sizeof(uint32_t)];

    //处理data，第4个字节之后为真实数据
    NSData *realData = [data subdataWithRange:NSMakeRange(headerLen, data.length - headerLen)];

    switch (messageType) {
        case SocketMessageType_StreamStart: {
            WSLog(@"======ws didReceiveData StreamStart");
            [self.dataStream open];
            self.handleReceiveData(messageType, nil);
            break;
        }
        case SocketMessageType_Streaming: {
            WSLog(@"======ws didReceiveData Streaming ...");
            NSUInteger dataLength = [realData length];
            NSInteger writeLen = [self.dataStream write:[realData bytes] maxLength:dataLength];
            if(dataLength > writeLen){  //发生错误
                self.streamFilePath = nil;
                self.streamError = [self.dataStream streamError];
                [self.dataStream close];
                self.dataStream = nil;
                return;
            }
            self.handleReceiveData(messageType, nil);
            break;
        }
        case SocketMessageType_StreamEnd: {
            WSLog(@"======ws didReceiveData StreamEnd");
            if(self.dataStream && self.dataStream.streamStatus != NSStreamStatusClosed){
                [self.dataStream close];
                self.dataStream = nil;
            }
            NSData *pathData = self.streamError==nil ?[self.streamFilePath dataUsingEncoding:NSUTF8StringEncoding] : nil;
            self.handleReceiveData(messageType, pathData);
            break;
        }
        default: {
            self.handleReceiveData(messageType, realData);
            break;
        }
    }
}

//websocket链接关闭
- (void)webSocketDidClose:(WebSocket *)ws{
    WSLog(@"=======%s",__FUNCTION__);
//    [self.webSocket stop];
//    self.webSocket = nil;
}



@end

File: LWWebSocket_swift/WebSocketManager.swift
//
//  WebSocketManager.swift
//  LWWebSocket
//
//  Created by Swift Version Conversion
//  Copyright © 2024. All rights reserved.
//

import Foundation

/// Main WebSocket server manager for in-app WebSocket communication
public class WebSocketManager: NSObject {

    // MARK: - Singleton

    public static let shared = WebSocketManager()

    private override init() {
        super.init()
    }

    // MARK: - Properties

    /// Callback for receiving text messages
    public var handleReceiveMessage: ((UInt32, String) -> Void)?

    /// Callback for receiving binary data
    public var handleReceiveData: ((UInt32, Data?) -> Void)?

    private var httpServer: HTTPServer?
    private weak var webSocket: MyWebSocket?

    // Stream properties for file transfer
    private var dataStream: OutputStream?
    private var streamFilePath: String?
    private var streamError: Error?

    // MARK: - Computed Properties

    private var generatedStreamFilePath: String {
        if streamFilePath == nil {
            streamFilePath = NSTemporaryDirectory() + UUID().uuidString
        }
        return streamFilePath!
    }

    private var outputStream: OutputStream {
        if dataStream == nil {
            dataStream = OutputStream(toFileAtPath: generatedStreamFilePath, append: true)
        }
        return dataStream!
    }

    var myURI: String {
        return "/service"
    }

    // MARK: - Server Management

    /// Start the WebSocket server on specified port
    /// - Parameters:
    ///   - port: Port number to listen on
    ///   - webPath: Web root directory path
    public func startServer(port: UInt16, webPath: String) {
        guard httpServer?.isRunning != true else {
            return
        }

        let server = HTTPServer()
        httpServer = server

        // Use custom HTTP connection class
        server.connectionClass = MyHTTPConnection.self

        // Broadcast presence via Bonjour
        server.setType("_http._tcp.")

        // Set port
        server.port = port

        // Setup document root
        let documentPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first ?? ""
        var absWebPath: String

        if webPath.hasPrefix("/var") {
            absWebPath = webPath
        } else {
            absWebPath = (documentPath as NSString).appendingPathComponent(webPath)
        }

        var isDirectory: ObjCBool = false
        let existsWebPath = FileManager.default.fileExists(atPath: absWebPath, isDirectory: &isDirectory)

        if !isDirectory.boolValue || !existsWebPath {
            do {
                try FileManager.default.createDirectory(atPath: absWebPath, withIntermediateDirectories: true, attributes: nil)
                WSLog("Successfully created directory at: %@", absWebPath)
            } catch {
                WSLog("Failed to create directory: %@", error.localizedDescription)
                absWebPath = documentPath
            }
        }

        server.documentRoot = absWebPath

        WSLog("=====webpath:%@", webPath)

        // Start the server
        do {
            try server.start()
        } catch {
            WSLog("start server error:%@", error.localizedDescription)
        }
    }

    /// Stop the WebSocket server
    public func stopServer() {
        if let server = httpServer, server.isRunning {
            server.stop()
        }
    }

    // MARK: - Sending Methods

    /// Send a text message through WebSocket
    /// - Parameter message: The message string to send
    /// - Returns: Success status
    @discardableResult
    public func sendMessage(_ message: String) -> Bool {
        guard let ws = webSocket else {
            return false
        }

        ws.sendMessage(message)
        return true
    }

    /// Send binary data through WebSocket
    /// - Parameter data: The data to send
    /// - Returns: Success status
    @discardableResult
    public func sendData(_ data: Data) -> Bool {
        guard let ws = webSocket else {
            return false
        }

        var sendData = constructData(withMessageType: .data)
        sendData.append(data)
        ws.send(sendData, isBinary: true)

        return true
    }

    /// Send file data through WebSocket using streaming
    /// - Parameter fileURL: URL of the file to send
    /// - Returns: Success status
    @discardableResult
    public func sendData(withFileURL fileURL: URL) -> Bool {
        guard let ws = webSocket else {
            return false
        }

        // Send stream start
        let startData = constructData(withMessageType: .streamStart)
        ws.send(startData, isBinary: true)

        // Read and send file in chunks
        do {
            let fileHandle = try FileHandle(forReadingFrom: fileURL)
            defer {
                if #available(iOS 13.0, *) {
                    try? fileHandle.close()
                } else {
                    fileHandle.closeFile()
                }
            }

            while true {
                let data: Data
                if #available(iOS 13.0, *) {
                    guard let chunk = try fileHandle.read(upToCount: 10240), !chunk.isEmpty else {
                        break
                    }
                    data = chunk
                } else {
                    data = fileHandle.readData(ofLength: 10240)
                    if data.isEmpty {
                        break
                    }
                }

                var streamingData = constructData(withMessageType: .streaming)
                streamingData.append(data)
                ws.send(streamingData, isBinary: true)
            }
        } catch {
            WSLog("===error:%@", error.localizedDescription)
        }

        // Send stream end
        let endData = constructData(withMessageType: .streamEnd)
        ws.send(endData, isBinary: true)

        return true
    }

    // MARK: - Helper Methods

    /// Construct data with message type header
    /// - Parameter messageType: The message type to prepend
    /// - Returns: Data with message type header
    private func constructData(withMessageType messageType: LWSocketMessageType) -> Data {
        var data = Data()
        var type = messageType.rawValue
        data.append(Data(bytes: &type, count: MemoryLayout<UInt32>.size))
        return data
    }

    /// Send initial hello message
    private func sendActiveBinaryData() {
        guard let ws = webSocket else { return }

        var data = Data()
        var messageType = LWSocketMessageType.hello.rawValue
        data.append(Data(bytes: &messageType, count: MemoryLayout<UInt32>.size))
        data.append("hello world!".data(using: .utf8)!)

        ws.send(data, isBinary: true)
    }
}

// MARK: - WebSocketDelegate

extension WebSocketManager: WebSocketDelegate {

    public func webSocketDidOpen(_ ws: WebSocket) {
        WSLog("=======%s", #function)

        if let myWs = ws as? MyWebSocket {
            webSocket = myWs
            myWs.startHeartBeatRecvTimer()
        }

        sendActiveBinaryData()
    }

    public func webSocket(_ ws: WebSocket, didReceiveMessage msg: String) {
        WSLog("=======%s", #function)

        guard !msg.isEmpty,
              let data = msg.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any] else {
            return
        }

        WSLog("=======didReceiveMessage msg:%@", msg)

        guard let messageTypeValue = dict["messageType"] as? Int else {
            return
        }

        let messageType = UInt32(messageTypeValue)

        // Handle heartbeat
        if messageType == LWSocketMessageType.heartBeat.rawValue,
           let myWs = ws as? MyWebSocket {
            myWs.startHeartBeatRecvTimer()
            return
        }

        // Handle regular message
        if let messageBody = dict["messageBody"] as? String {
            handleReceiveMessage?(messageType, messageBody)
        }
    }

    public func webSocket(_ ws: WebSocket, didReceiveData data: Data) {
        WSLog("=======%s", #function)

        guard handleReceiveData != nil else {
            return
        }

        let headerLen = 4
        guard data.count >= headerLen else {
            return // Invalid message
        }

        // Extract message type from first 4 bytes
        let headerData = data.subdata(in: 0..<headerLen)
        let messageType = headerData.withUnsafeBytes { $0.load(as: UInt32.self) }

        // Extract real data (after header)
        let realData = data.subdata(in: headerLen..<data.count)

        switch LWSocketMessageType(rawValue: messageType) {
        case .streamStart:
            WSLog("======ws didReceiveData StreamStart")
            outputStream.open()
            handleReceiveData?(messageType, nil)

        case .streaming:
            WSLog("======ws didReceiveData Streaming ...")
            let dataLength = realData.count
            let writeLen = realData.withUnsafeBytes { (bytes: UnsafeRawBufferPointer) -> Int in
                guard let baseAddress = bytes.baseAddress else { return 0 }
                return outputStream.write(baseAddress.assumingMemoryBound(to: UInt8.self), maxLength: dataLength)
            }

            if dataLength > writeLen {
                // Error occurred
                streamFilePath = nil
                streamError = outputStream.streamError
                outputStream.close()
                dataStream = nil
                return
            }
            handleReceiveData?(messageType, nil)

        case .streamEnd:
            WSLog("======ws didReceiveData StreamEnd")
            if let stream = dataStream, stream.streamStatus != .closed {
                stream.close()
                dataStream = nil
            }

            let pathData = streamError == nil ? streamFilePath?.data(using: .utf8) : nil
            handleReceiveData?(messageType, pathData)

        default:
            handleReceiveData?(messageType, realData)
        }
    }

    public func webSocketDidClose(_ ws: WebSocket) {
        WSLog("=======%s", #function)
    }
}
